File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/evaluation/GeomagLocationEvaluator.java
package se.gustavkarlsson.aurora_notifier.android.evaluation;

import javax.inject.Inject;

import se.gustavkarlsson.aurora_notifier.android.models.factors.GeomagLocation;

public class GeomagLocationEvaluator implements ChanceEvaluator<GeomagLocation> {

	@Inject
	GeomagLocationEvaluator() {
	}

	public Chance evaluate(GeomagLocation geomagLocation) {
		Float latitude = geomagLocation.getLatitude();
		if (latitude == null) {
			return Chance.unknown();
		}
		double absoluteLatitude = Math.abs(latitude);
		double chance = (1.0 / 12.0) * absoluteLatitude - (55.0 / 12.0); // 55-67
		if (chance > 1.0) {
			chance = 2.0 - chance;
		}
		return Chance.of(chance);
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/services_impl/evaluation/GeomagLocationEvaluator.kt
package se.gustavkarlsson.skylight.android.services_impl.evaluation

import se.gustavkarlsson.skylight.android.entities.Chance
import se.gustavkarlsson.skylight.android.entities.GeomagLocation
import se.gustavkarlsson.skylight.android.services.ChanceEvaluator

object GeomagLocationEvaluator : ChanceEvaluator<GeomagLocation> {

    override fun evaluate(value: GeomagLocation): Chance {
        val latitude = value.latitude
        val absoluteLatitude = Math.abs(latitude)
        var chance = ((1.0 / FLEX) * absoluteLatitude) - ((BEST - FLEX) / FLEX)
        if (chance > 1.0) {
            chance = 2.0 - chance
        }
        return Chance(chance)
    }

	const val BEST = 67.0
	const val FLEX = 13.0
}
